### DISCUSSION TITLE
NATS stopped to handle new messages after many NATS.Client.NATSSlowConsumerException

### DISCUSSION THREAD

**Discussion by esolCrusador** [nats-io/nats.net.v1#533] [Q&A] [answered]
NATS.Client.NATSSlowConsumerException at STAN.Client.Connection.PublishAsync
NATS.Client.NATSSlowConsumerException
NATS.Client.NATSSlowConsumerException at NATS.Client.Connection.publish

We are using STAN.Client and there we are using PublishAsync, Ack methods.
Project is EventSourcing based reporting. We are handling different events of user session. 
When we are handling message we first are storing event to datatabase and when we know that session is done we are sending PublihAsync() command to calculate session. 
Than our session calculation was slown down to 30+ seconds. 
And we have 2 subscriptions for this service 
	**Handle** Event -> Store -> **PublishAsync()** UpdateServiceSessionsCommand
	**Handle** UpdateServiceSessionsCommand -> Store (30 sec timeout).
At some moment we started to receive NATSSlowConsumerException looks like there were too many messages in client.
We had NATSSlowConsumerException for both Ack() and PublishAsync().
And after some time both Ack() and PublishAsync() stopped to work.
In the end client just stopped to handle any messages. It happened for all nodes of this service. And only restart helps to start them message handling again.

I think in this case when there are too many NATS client messages it just should not consume new messages and consume them only when his Ack() and PublishAsync() are flushed. Because otherwise it leads to something like deadlock.

**Comment by scottf** [comment, marked as answer]
@esolCrusador Are you working on a new project? If so you should definitely move to JetStream. There is a feature called flow control that will recognize when you are taking too long to handle messages and stop pushing them to you until the flow control message is acknowledged by the client, which requires you processing all the messages before the flow control.
If you are on STAN and staying I'm not sure how to help you, STAN is not being supported anymore.

Moving this to the discussion area instead if issues.
